File: backend/src/main/java/com/boxoffice/service/CategoryService.java
/*
 * myRC - Category Service Interface
 * Copyright (c) 2026 myRC Team
 * Licensed under MIT License
 *
 * Author: myRC Team
 * Date: 2026-01-27
 * Version: 1.0.0
 *
 * Description:
 * Service interface for Category management operations.
 */
package com.boxoffice.service;

import com.boxoffice.dto.CategoryDTO;
import java.util.List;
import java.util.Optional;

/**
 * Service interface for Category management operations.
 * Categories are used for grouping both funding and spending items.
 *
 * @author myRC Team
 * @version 1.0.0
 * @since 2026-01-27
 */
public interface CategoryService {

  /**
   * Get all categories for a fiscal year.
   *
   * @param fiscalYearId the fiscal year ID
   * @param username the requesting user's username
   * @return list of categories
   * @throws IllegalArgumentException if user doesn't have access
   */
  List<CategoryDTO> getCategoriesByFiscalYearId(Long fiscalYearId, String username);

  /**
   * Get a specific category by ID.
   *
   * @param categoryId the category ID
   * @param username the requesting user's username
   * @return optional category
   */
  Optional<CategoryDTO> getCategoryById(Long categoryId, String username);

  /**
   * Create a new category for a fiscal year.
   * Only custom categories can be created - default categories are system-managed.
   *
   * @param fiscalYearId the fiscal year ID
   * @param username the requesting user's username
   * @param name the category name
   * @param description the category description
   * @return the created category
   * @throws IllegalArgumentException if user doesn't have write access or category name already exists
   */
  CategoryDTO createCategory(Long fiscalYearId, String username, String name, String description);

  /**
   * Update an existing category.
   * Only custom categories can be updated - default categories are read-only.
   *
   * @param categoryId the category ID
   * @param username the requesting user's username
   * @param name the new category name
   * @param description the new category description
   * @return the updated category
   * @throws IllegalArgumentException if user doesn't have write access or category is a default category
   */
  CategoryDTO updateCategory(Long categoryId, String username, String name, String description);

  /**
   * Delete a category.
   * Only custom categories can be deleted - default categories cannot be deleted.
   *
   * @param categoryId the category ID
   * @param username the requesting user's username
   * @throws IllegalArgumentException if user doesn't have write access or category is a default category
   */
  void deleteCategory(Long categoryId, String username);

  /**
   * Ensure default categories exist for a fiscal year.
   * Creates the default categories if they don't exist:
   * - Compute
   * - GPUs
   * - Storage
   * - Software Licenses
   * - Small Procurement
   * - Contractors
   *
   * @param fiscalYearId the fiscal year ID
   * @param username the requesting user's username
   * @return list of all categories including defaults
   */
  List<CategoryDTO> ensureDefaultCategoriesExist(Long fiscalYearId, String username);

  /**
   * Reorder categories.
   *
   * @param fiscalYearId the fiscal year ID
   * @param username the requesting user's username
   * @param categoryIds the list of category IDs in the new order
   * @return the reordered categories
   * @throws IllegalArgumentException if user doesn't have write access
   */
  List<CategoryDTO> reorderCategories(Long fiscalYearId, String username, List<Long> categoryIds);

  /**
   * Initialize default categories for a fiscal year (internal use).
   * This method is called during FY creation and does not require user authentication.
   *
   * @param fiscalYearId the fiscal year ID
   */
  void initializeDefaultCategories(Long fiscalYearId);
}

File: frontend/src/app/services/category.service.ts
/**
 * Category Service for myRC application.
 * Handles all API communication for category management.
 *
 * @author myRC Team
 * @version 1.0.0
 * @since 2026-01-27
 * @license MIT
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { Category, FundingType } from '../models/category.model';

/**
 * Request body for creating a category.
 */
export interface CategoryCreateRequest {
  name: string;
  description?: string;
  fundingType?: FundingType;
}

/**
 * Request body for updating a category.
 */
export interface CategoryUpdateRequest {
  name?: string;
  description?: string;
  fundingType?: FundingType;
}

/**
 * Service for managing categories within fiscal years.
 * Categories are used to group both funding and spending items.
 * Default categories are read-only; only custom categories can be added, renamed, and deleted.
 */
@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  private baseUrl(rcId: number, fyId: number): string {
    return `/api/responsibility-centres/${rcId}/fiscal-years/${fyId}/categories`;
  }

  constructor(private http: HttpClient) {}

  /**
   * Get all categories for a fiscal year.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @returns Observable of categories array
   */
  getCategoriesByFY(rcId: number, fyId: number): Observable<Category[]> {
    return this.http.get<Category[]>(this.baseUrl(rcId, fyId), { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  /**
   * Get a specific category by ID.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @param categoryId The category ID
   * @returns Observable of the category
   */
  getCategory(rcId: number, fyId: number, categoryId: number): Observable<Category> {
    return this.http.get<Category>(`${this.baseUrl(rcId, fyId)}/${categoryId}`, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  /**
   * Create a new category for a fiscal year.
   * Note: Only custom categories can be created. Default categories are system-managed.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @param request The category creation request
   * @returns Observable of the created category
   */
  createCategory(rcId: number, fyId: number, request: CategoryCreateRequest): Observable<Category> {
    return this.http.post<Category>(this.baseUrl(rcId, fyId), request, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  /**
   * Update an existing category.
   * Note: Only custom categories can be updated. Default categories are read-only.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @param categoryId The category ID
   * @param request The category update request
   * @returns Observable of the updated category
   */
  updateCategory(rcId: number, fyId: number, categoryId: number, request: CategoryUpdateRequest): Observable<Category> {
    return this.http.put<Category>(`${this.baseUrl(rcId, fyId)}/${categoryId}`, request, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  /**
   * Delete a category.
   * Note: Only custom categories can be deleted. Default categories cannot be deleted.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @param categoryId The category ID
   * @returns Observable of void
   */
  deleteCategory(rcId: number, fyId: number, categoryId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl(rcId, fyId)}/${categoryId}`, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  /**
   * Ensure default categories exist for a fiscal year.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @returns Observable of categories array including defaults
   */
  ensureDefaults(rcId: number, fyId: number): Observable<Category[]> {
    return this.http.post<Category[]>(`${this.baseUrl(rcId, fyId)}/ensure-defaults`, {}, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  /**
   * Reorder categories.
   *
   * @param rcId The responsibility centre ID
   * @param fyId The fiscal year ID
   * @param categoryIds Array of category IDs in the new order
   * @returns Observable of reordered categories array
   */
  reorderCategories(rcId: number, fyId: number, categoryIds: number[]): Observable<Category[]> {
    return this.http.post<Category[]>(
      `${this.baseUrl(rcId, fyId)}/reorder`,
      { categoryIds },
      { withCredentials: true }
    ).pipe(catchError(this.handleError));
  }

  /**
   * Handle HTTP errors.
   */
  private handleError(error: HttpErrorResponse): Observable<never> {
    let errorMessage = 'An unexpected error occurred';
    
    if (error.error instanceof ErrorEvent) {
      // Client-side error
      errorMessage = error.error.message;
    } else {
      // Server-side error
      if (error.status === 401) {
        errorMessage = 'Authentication required. Please log in.';
      } else if (error.status === 403) {
        if (error.error?.message) {
          errorMessage = error.error.message;
        } else {
          errorMessage = 'You do not have permission to perform this action.';
        }
      } else if (error.status === 404) {
        errorMessage = 'Category not found.';
      } else if (error.status === 409) {
        errorMessage = 'A category with this name already exists for this fiscal year.';
      } else if (error.error?.message) {
        errorMessage = error.error.message;
      }
    }
    
    return throwError(() => new Error(errorMessage));
  }
}
